### ISSUE TITLE
Wrong SyntaxWalker visiting order for RegionDirectiveTriviaSyntax

### ISSUE THREAD

**Issue by meancrazy** [dotnet/roslyn#14324]
**Version Used**: 
1.3.2 or 2.0.0-beta3
**Steps to Reproduce**:

1.  copy-paste and run following code:
```csharp
using System;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Test
{
    class Program
    {

        static void Main(string[] args)
        {

            const string data = @"
#region r0
    namespace ns0 { class c0 { } }
#endregion r0
#region r1
    namespace ns1 { class c1 { } }
#endregion r1
";

            var tree = CSharpSyntaxTree.ParseText(data);
            new TestSyntaxWalker().Visit(tree.GetRoot());
            Console.ReadLine();
        }
    }

    public class TestSyntaxWalker : CSharpSyntaxWalker
    {
        public TestSyntaxWalker() : base(SyntaxWalkerDepth.StructuredTrivia) { }

        public override void VisitNamespaceDeclaration(NamespaceDeclarationSyntax node)
        {
            Console.WriteLine($"VisitNamespaceDeclaration\t{((IdentifierNameSyntax)node.Name).Identifier.Text}");
            base.VisitNamespaceDeclaration(node);
        }

        public override void VisitRegionDirectiveTrivia(RegionDirectiveTriviaSyntax node)
        {
            Console.WriteLine($"VisitRegionDirectiveTrivia\t{node}");
            base.VisitRegionDirectiveTrivia(node);
        }

        public override void VisitEndRegionDirectiveTrivia(EndRegionDirectiveTriviaSyntax node)
        {
            Console.WriteLine($"VisitEndRegionDirectiveTrivia\t{node}");
            base.VisitEndRegionDirectiveTrivia(node);
        }

        public override void VisitClassDeclaration(ClassDeclarationSyntax node)
        {
            Console.WriteLine($"VisitClassDeclaration\t\t{node.Identifier.Text}");
            base.VisitClassDeclaration(node);
        }
    }
}
```
2. See program output

**Expected Behavior**:
Regions declared before namespace or class must be visited early.
Program output is:
```
VisitRegionDirectiveTrivia      #region r0
VisitNamespaceDeclaration       ns0
VisitClassDeclaration           c0
VisitEndRegionDirectiveTrivia   #endregion r0
VisitRegionDirectiveTrivia      #region r1
VisitNamespaceDeclaration       ns1
VisitClassDeclaration           c1
VisitEndRegionDirectiveTrivia   #endregion r1
```
**Actual Behavior**:
Walker is visiting namespace or class declaration and then returning to region and endregion trivia.
Program output is:
```
VisitNamespaceDeclaration       ns0
VisitRegionDirectiveTrivia      #region r0
VisitClassDeclaration           c0
VisitNamespaceDeclaration       ns1
VisitEndRegionDirectiveTrivia   #endregion r0
VisitRegionDirectiveTrivia      #region r1
VisitClassDeclaration           c1
VisitEndRegionDirectiveTrivia   #endregion r1
```

**Comment by CyrusNajmabadi** [comment]
This behavior is by design.  The walker walks the tree from the root downwards.  It can't get to the preprocessor regions without going through the nodes they're on first.  i.e. in order to hit "#region r0", it has to walk into your ns0 namespace.

**Comment by meancrazy** [comment]
Ok, if I change data to 
```
namespace ns0 { 
    #region r0
        class c0 { }
    #endregion r0
}
```
and run it again I getting
```
VisitNamespaceDeclaration       ns0
VisitClassDeclaration           c0
VisitRegionDirectiveTrivia      #region r0
VisitEndRegionDirectiveTrivia   #endregion r0
```
but as you say walker already have namespace node and can walk to region trivia. What's wrong?

**Comment by CyrusNajmabadi** [comment]
Nothing is wrong.  You hit the namespace node, so you printed ns0.  Then you called base.VisitNamespaceNode().  That makes the walker visit all the parts of the namespace.  That means visiting "namespace" "ns0" "{" and then the member list.  When it visits the member list, it visits the class node.  When it hits that you printed out "c0".  YOu then called base.VisitClassDeclaration.  

That visited the "class" token. When the walker visited the "class" token, its default behavior was to walk into the trivia for that token.  When it walked into its trivia you hit the "#region".   After that, we then walked into "c0" "{" and "}".  Then we popped out and finished walking the namespace.  Specifically, the walker then hit hte final "}".  It then walked the trivia of that close curly and ran into the "#endregion".
